subprojects/platform-play/src/integTest/groovy/org/gradle/play/integtest/continuous/PlayReloadIntegrationTest.groovy

```diff
@@ -196,11 +204,130 @@ task otherTask {
         appIsRunningAndDeployed()
 
         when:
-        addHelloWorld()
+        addNewRoute()
 
         then:
         fails()
         !executedTasks.contains('runPlayBinary')
         errorPageHasTaskFailure("otherTask")
     }
+
+    void addPendingChangesHook() {
+        buildFile << """
+            ext.pendingChanges = new java.util.concurrent.atomic.AtomicBoolean(false)
+
+            void addPendingChangeListener() {
+                def pendingChangesManager = gradle.services.get(${PendingChangesManager.canonicalName})
+                pendingChangesManager.addListener new ${SingleFirePendingChangesListener.canonicalName}({
+                    synchronized(pendingChanges) {
+                        println "Pending changes detected"
+                        pendingChanges.set(true)
+                        pendingChanges.notifyAll()
+                    }
+                } as ${PendingChangesListener.canonicalName})
+            }
+
+            void waitForPendingChanges() {
+                synchronized(pendingChanges) {
+                    while(!pendingChanges.get()) {
+                        pendingChanges.wait()
+                    }
+                }
+            }
+        """
+    }
+
+    def "wait for changes to be built when a request comes in during initial app startup and there are pending changes"() {
+        addPendingChangesHook()
+        // Prebuild so we don't timeout waiting for the 'rebuild' trigger
+        executer.withTasks("playBinary").run()
+
+        server.start()
+        buildFile << """
+            tasks.withType(${PlayRun.name}) {
+                doLast {
+                    def routes = file('conf/routes').text
+                    if (!routes.contains("hello")) {
+                        addPendingChangeListener()
+                        // signal we're ready to modify
+                        ${server.callFromBuild("rebuild")}
+                        // test should have added the hello route, so wait until Gradle
+                        // detects this as a pending change
+                        waitForPendingChanges()
+                    }
+                }
+            }
+        """
+
+        when:
+        start("runPlayBinary")
+        def rebuild = server.expectAndBlock("rebuild")
+        rebuild.waitForAllPendingCalls()
+        // Trigger a change during the build
+        addNewRoute()
+        rebuild.releaseAll()
+        then:
+        appIsRunningAndDeployed()
+        runningApp.playUrl('hello').text == 'hello world'
+    }
+
+    def "wait for pending changes to be built if a request comes in during a build and there are pending changes"() {
+        server.start()
+        addPendingChangesHook()
+        buildFile << """
+            tasks.withType(${PlayRun.name}) {
```

Comment:
Adding this to `doLast` of the `PlayRun` task doesn't seem right to me. 

A. Test that browser request blocks until build is completed:
1) We need to execute `playRun`, and wait for the app to be ready. Perhaps we verify the response to make sure it's all good.
2) We then make a change that triggers a build. 
3) New build starts and a task _before_ this build finishes (perhaps when compiling the change).  Test code waits for this block, to be sure the change was delivered.
4) We make a request to the app. This should block, since there's a build in progress.
5) Release the block in the task: the app request should eventually return, containing changes from 2).

Since we'll be doing work concurrently with making the request, we'll need to use something like `ConcurrentTestUtil`. But we shouldn't need a 'PendingChangesListener` for this test case.

B. Test that browser request blocks until second build is completed, when a change is made during first build _before_ the request.
1&2) Same as above
3) New build starts and a task registers a pending changes listener before blocking. Test code waits for this block.
4) Make another change to input file, wait for pending changes listener to fire. The pending changes listener could make a different blocking call that we can wait for in the test.
5) Once the change was delivered, make an app request: should block because the _first_ build is still in progress.
6) Just wait for the app request to return: should contain the _second_ changes.
7) Verify that 2 builds executed.

C. Test that browser request blocks until second build is completed, when a change is made during first build _after_ the request.
This is the same as B, except we'd make the app request before making the second input file change and waiting for the pending change to be delivered.

This stuff is complex, but I think we could make it simpler by modelling the concurrency between the app request and the file changes.